### PR TITLE
The return of the xcm

### PR DESCRIPTION
**Pull request by shaunxw**

### CODE REVIEW

File: xtokens/Cargo.toml [hunk 1]
```diff
@@ -20,23 +20,23 @@ frame-system = { git = "https://github.com/paritytech/substrate", branch = "polk
...
-pallet-balances = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.4" }
-xcm-simulator = { git = "https://github.com/shaunxw/xcm-simulator", branch = "master" }
-cumulus-pallet-xcm-handler = { git = "https://github.com/paritytech/cumulus", branch = "polkadot-v0.9.4" }
-parachain-info = { git = "https://github.com/paritytech/cumulus", branch = "polkadot-v0.9.4" }
-xcm-executor = { git = "https://github.com/paritytech/polkadot", branch = "release-v0.9.4" }
-xcm-builder = { git = "https://github.com/paritytech/polkadot", branch = "release-v0.9.4" }
-
-orml-tokens = { path = "../tokens", version = "0.4.1-dev" }
-orml-traits = { path = "../traits", version = "0.4.1-dev" }
+# sp-core = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.4" }
```
Comment (xlc) [inline xtokens/Cargo.toml:29]: why this?

File: xtokens/src/lib.rs [hunk 2]
```diff
@@ -110,18 +130,18 @@ pub mod module {
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {
 		/// Transfer native currencies.
-		#[transactional]
-		#[pallet::weight(1000)]
+		#[pallet::weight(Pallet::<T>::weight_of_transfer(currency_id.clone(),*amount, &dest))]
```
Comment (xlc) [inline xtokens/src/lib.rs:133]:
```suggestion
		#[pallet::weight(Pallet::<T>::weight_of_transfer(*currency_id, *amount, &dest))]
```

File: xtokens/Cargo.toml (hunk 1 above)
Comment (shaunxw) [reply xtokens/Cargo.toml:29]: I commented out all unit tests related things and will bring them back in another PR, after `xcm-simulator` is updated.

File: xtokens/src/lib.rs (hunk 2 above)
Comment (shaunxw) [reply xtokens/src/lib.rs:133]: `currency_id` can't be moved here. I'll update the format.

File: xtokens/src/lib.rs [hunk 3]
```diff
@@ -22,28 +22,37 @@
...
+use xcm::v0::prelude::*;
+use xcm_executor::traits::WeightBounds;
 
+pub use module::*;
 use orml_traits::location::{Parse, Reserve};
-use orml_xcm_support::XcmHandler;
 
-mod mock;
-mod tests;
+// mod mock;
```
Comment (xlc) [inline xtokens/src/lib.rs:44]: is this intended?

Comment (shaunxw) [reply xtokens/src/lib.rs:44]: Yes. A lot of work is needed on `xcm-simulator` to update mock and unit tests. It will be done in another PR.

**Review by xlc — APPROVED** [review]

File: xcm-support/src/lib.rs [hunk 4]
```diff
@@ -25,8 +25,8 @@ mod currency_adapter;
 mod tests;
 
 /// The XCM handler to execute XCM locally.
-pub trait XcmHandler<AccountId> {
-	fn execute_xcm(origin: AccountId, xcm: Xcm) -> DispatchResult;
+pub trait ExecuteXcm<AccountId, Call> {
+	fn execute_xcm(origin: AccountId, xcm: Xcm<Call>) -> DispatchResult;
 }
```
Comment (apopiak) [inline xcm-support/src/lib.rs:30]: Why have your own trait for this?

File: xtokens/src/lib.rs [hunk 5]
```diff
@@ -255,13 +312,100 @@ pub mod module {
...
+							dest,
+							effects: vec![],
+						}],
+					},
+					ToReserve | ToNonReserve => {
+						WithdrawAsset {
+							assets: vec![asset.clone()],
+							effects: vec![InitiateReserveWithdraw {
+								assets: vec![All],
+								// `dest` is always (equal to) `reserve` in both cases
```
Comment (apopiak) [inline xtokens/src/lib.rs:372]: That seems wrong?

File: xtokens/src/lib.rs [hunk 6]
```diff
@@ -255,13 +312,100 @@ pub mod module {
...
+		/// Returns weight of `transfer` call.
+		fn weight_of_transfer(currency_id: T::CurrencyId, amount: T::Balance, dest: &MultiLocation) -> Weight {
+			if let Some(id) = T::CurrencyIdConvert::convert(currency_id) {
+				let asset = MultiAsset::ConcreteFungible {
+					id,
+					amount: amount.into(),
+				};
+				Self::weight_of_transfer_multiasset(&asset, &dest)
+			} else {
+				0
```
Comment (apopiak) [inline xtokens/src/lib.rs:394]: You'll likely want to benchmark to determine a more accurate weight here.

File: xcm-support/src/lib.rs (hunk 4 above)
Comment (shaunxw) [reply xcm-support/src/lib.rs:30]: It was for the previous version of xtokens, will remove it.

File: xtokens/src/lib.rs (hunk 5 above)
Comment (shaunxw) [reply xtokens/src/lib.rs:372]: I think in both cases (non self reserve assets) the XCM will be sent to reserve chain to update reserve accounts firstly. If dest is reserve chain, the XCM execution stops there, or the reserve chain will send another XCM to dest notifying the reserve account change.

**Pull request merged**